src/main/java/build/buildfarm/worker/ReportResultStage.java

```diff
@@ -0,0 +1,182 @@
+    final String operationName = operationContext.operation.getName();
+    Poller poller = new Poller(worker.config.getOperationPollPeriod(), () -> {
+          boolean success = worker.instance.pollOperation(
+              operationName,
+              ExecuteOperationMetadata.Stage.EXECUTING);
+          return success;
+        });
+    new Thread(poller).start();
+
+    ActionResult.Builder resultBuilder;
+    try {
+      resultBuilder = operationContext
+          .operation.getResponse().unpack(ExecuteResponse.class).getResult().toBuilder();
+    } catch (InvalidProtocolBufferException ex) {
+      poller.stop();
+      return null;
+    }
+
+    ImmutableList.Builder<ByteString> contents = new ImmutableList.Builder<>();
+    if (resultBuilder.getExitCode() == 0) {
+      CASInsertionControl control = worker.config.getFileCasControl();
+      for (String outputFile : operationContext.action.getOutputFilesList()) {
+        Path outputPath = operationContext.execDir.resolve(outputFile);
+        if (!Files.exists(outputPath)) {
+          continue;
+        }
+
+        // FIXME put the output into the fileCache
+        // FIXME this needs to be streamed to the server, not read to completion, but
+        // this is a constraint of not knowing the hash, however, if we put the entry
+        // into the cache, we can likely do so, stream the output up, and be done
+        //
+        // will run into issues if we end up blocking on the cache insertion, might
+        // want to decrement input references *before* this to ensure that we cannot
+        // cause an internal deadlock
+
+        ByteString content;
+        try {
+          InputStream inputStream = Files.newInputStream(outputPath);
+          content = ByteString.readFrom(inputStream);
+          inputStream.close();
+        } catch (IOException ex) {
+          continue;
+        }
+        OutputFile.Builder outputFileBuilder = resultBuilder.addOutputFilesBuilder()
+            .setPath(outputFile)
+            .setIsExecutable(Files.isExecutable(outputPath));
+        boolean withinLimit = content.size() <= worker.config.getFileCasControl().getLimit();
+        if (withinLimit) {
+          outputFileBuilder.setContent(content);
+        }
+        if (control.getPolicy() == CASInsertionControl.Policy.ALWAYS_INSERT ||
+            (!withinLimit && control.getPolicy() == CASInsertionControl.Policy.INSERT_ABOVE_LIMIT)) {
+          contents.add(content);
+
+          // FIXME make this happen with putAllBlobs
+          Digest outputDigest = Digests.computeDigest(content);
+          outputFileBuilder.setDigest(outputDigest);
+        }
+      }
+
+      try {
+        worker.instance.putAllBlobs(contents.build());
+      } catch (IOException ex) {
+      } catch (InterruptedException ex) {
+        poller.stop();
+        return null;
+      }
+    }
+
+    ActionResult result = resultBuilder.build();
+    if (!operationContext.action.getDoNotCache()) {
```

Comment:
and if exit code is 0, right? You don't want to put failed actions into the cache.

Reply:
Likewise added this test. Mentioned that we may want to make this configurable, in cases where we want to trust the failure.  The bazel-supported configuration will only have this happen on success, correct.